File: get-quotes.pl
# Copyright 2018 Tomas Brabec
#
# See LICENSE for license details.

# -----------------------------------------------------------------------------
# Description:
#   Quotes stock and currency symbols from DB.
#
#   The list of stocks and currencies is taken from DB. Currencies are
#   quoted only when the base currency is given (`-b` option).
#
#   Quotes are obtained from the following sources with decreasing priority:
#   - Cached quotes from DB
#   - Yahoo Finance
#   - AlphaVantage.
# -----------------------------------------------------------------------------
use DBI;
use Getopt::Long::Descriptive;
#use Switch;
#use Data::Dumper;
use Finance::Quote;
use Array::Utils;
use xfrs;
use strict;
use warnings;


my ($opt, $usage) = describe_options(
  '%c %o',
  [ 'db|d=s',    "Sqlite3 DB file to import into",   { default  => "xfrs.sqlite3.db" } ],
  [ 'base|b=s',  "base currency",   { default  => "" } ],
  [],
  [ 'help|h',       "print usage message and exit", { shortcircuit => 1 } ],
);


my $dbh = DBI->connect("dbi:SQLite:dbname=".$opt->db,"","",{ RaiseError => 1 }) or die $DBI::errstr;;

my @stocks = xfrs::getStocks($dbh);
my @currencies = xfrs::getCurrencies($dbh);


my $q = Finance::Quote->new;

$q->timeout(30);
my @attrs = ("last","currency");
my %quotes;

# Remaps attributes from the info structure returned by a quote.
my %attrMap = (
    'last' => 'price',
    'currency' => 'currency'
);


# Quote stocks
# ------------
# obtain cached quotes
my %qtCacheStocks = xfrs::getCachedQuote($dbh,'',@stocks);
foreach my $s (keys %qtCacheStocks) {
    $quotes{$s} = $qtCacheStocks{$s};
    print "$s (cache) ....\n";
    foreach my $a (@attrs) {
        print "\t$s.$a = ".$quotes{$s}->{$attrMap{$a}}."\n";
    }
}

# obtain additional quotes (if needed) from
# - Yahoo finance
# - AlphaVantage
foreach my $qtSrc ('yahoo_json', 'alphavantage') {
    if (scalar @stocks > scalar keys %quotes) {
        my @syms;
        foreach my $s (@stocks) {
            push(@syms,$s) unless (exists($quotes{$s}));
        }
    
        my %qs = $q->fetch($qtSrc,@syms);
        foreach my $s (@syms) {
            next unless (exists($qs{$s,'success'}) && $qs{$s,'success'} == 1);
    
            print "$s ($qtSrc) ....\n";
            foreach my $a (@attrs) {
                if (exists($qs{$s,$a})) {
                    $quotes{$s}->{$attrMap{$a}} = $qs{$s,$a};
                    print "\t$s.$a = ".$qs{$s,$a}."\n";
                }
            }
        }
    }
}

# report failed stock quotes
if (scalar @stocks > scalar keys %quotes) {
    foreach my $s (@stocks) {
        print "$s ... failed\n" unless (exists($quotes{$s}));
    }
}

# Quote currencies
# ----------------
my %curconv;
if ($opt->base ne "") {
    my %curQuotes;
    my @curs = @currencies;
    my $base = $opt->base;

    # add currencies of the stocks
    foreach my $s (keys %quotes) {
        my $rec = $quotes{$s};
        if (exists $rec->{'currency'}) {
            push(@curs,$rec->{'currency'});
        }
    }
    @curs = Array::Utils::unique(@curs);

    # get cached quotes first
    my @cacheSyms;
    foreach my $c (@curs) {
        push(@cacheSyms,$c.$base);
    }
    my %qtCacheCurs = xfrs::getCachedQuote($dbh,'',@cacheSyms);
    foreach my $c (@curs) {
        if (exists($qtCacheCurs{$c.$base})) {
            $curQuotes{$c} = $qtCacheCurs{$c.$base};
            print "$c (cache) ....\n";
            foreach my $a (@attrs) {
                print "\t$c.$a = ".$curQuotes{$c}->{$attrMap{$a}}."\n";
            }
        }
    }

    # append the base currency (if needed)
    if (!exists($curQuotes{$base})) {
        $curQuotes{$base} = {
            'price' => 1.0,
            'currency' => $base
        };
    }

    # quote conversion rates at Yahoo Finance (if needed)
    if (scalar @curs > scalar keys %curQuotes) {
        my %syms;
        foreach my $s (@curs) {
            $syms{$s}=$s.$base."=X" unless (exists($curQuotes{$s}));
        }
   
        my %qs = $q->fetch("yahoo_json",values %syms);
#print Dumper(\%qs);
        foreach my $s (keys %syms) {
            next unless (exists($qs{$syms{$s},'success'}) && $qs{$syms{$s},'success'} == 1);
    
            print "$s (yahoo_json) ....\n";
            foreach my $a (@attrs) {
                if (exists($qs{$syms{$s},$a})) {
                    $curQuotes{$s}->{$attrMap{$a}} = $qs{$syms{$s},$a};
                    print "\t$s.$a = ".$qs{$syms{$s},$a}."\n";
                }
            }
        }
    }

    # quote conversion rates (if needed) using the default API (AlphaVantage as of Finance::Quote 1.47)
    if (scalar @curs > scalar keys %curQuotes) {
        my @syms;
        foreach my $s (@curs) {
            push(@syms,$s) unless (exists($curQuotes{$s}));
        }

        foreach my $c (@syms) {
            if ($base eq $c) {
                $curQuotes{$c} = {
                    'price' => 1.0,
                    'currency' => $c
                };
            } else {
                my $convrate = $q->currency($c,$base);
                if (! defined $convrate ) { next; }
                $curQuotes{$c} = {
                    'price' => $convrate,
                    'currency' => $base
                };
            }
            print "$c/$base (defualt) ....\n";
            foreach my $a (@attrs) {
                print "\t$c.$a = ".$curQuotes{$c}->{$attrMap{$a}}."\n";
            }
        }
    }

    # populate the conversion rate hash
    foreach my $c (keys %curQuotes) {
        $curconv{$c} = $curQuotes{$c}->{'price'};
    }
}

# Print balance
# --------------
# get the actual balance
my %balance;
xfrs::getBalance($dbh, \%balance);

# collect NAV (net asset value)
my $nav = 0;

# print the cash balance
foreach my $c (@currencies) {
    print "$c,$balance{$c},$c";
    if (exists($curconv{$c})) {
        my $v = $balance{$c}*$curconv{$c};
        print ",".$v.",".$opt->base;
        $nav += $v;
    }
    print "\n";
}

# print the equity balance
foreach my $s (@stocks) {
    my $p = $quotes{$s}->{'price'} || 0;
    my $c = $quotes{$s}->{'currency'} || "";
    print "$s,".($balance{$s}*$p).",$c";
    if (exists $curconv{$c} && defined $curconv{$c}) {
        my $v = $balance{$s}*$p*$curconv{$c};
        print ",".$v.",".$opt->base;
        $nav += $v;
    }
    print "\n";
}

print "\nnav = $nav".$opt->base."\n";

#my $conversion_rate = $q->currency("AUD","USD");
#$q->set_currency("EUR");  # Return all info in Euros.
#
#$q->require_labels(qw/price date high low volume/);
#
#$q->failover(1); # Set failover support (on by default).
#
#my @stocks = ["RGR", "AOBC"];
#
#my %quotes  = $q->fetch("nasdaq",@stocks);
##my $hashref = $q->fetch("nyse",@stocks);
#
#foreach my $k (keys %quotes) {
#    print "$k = $quotes{$k}\n";
#}

File: get-quotes.py
import sqlite3
import sys
import xfrs
import argparse
import datetime
import re

# The list of stocks and currencies is given as an argument or
# else is taken from DB. Currencies are quoted only when
# the base currency is given (`-b` option).

parser = argparse.ArgumentParser(
        description="Quotes stock and currency symbols."
        )

# `db` option: Path to the XFRS database.
parser.add_argument('-d', '--db', type=str,
        default='xfrs.sqlite3.db', dest='db',
        help="Path to the DB file.",
        )

# `base` option: Symbol of the base currency.
parser.add_argument('-b', '--base', type=str,
        default=None, dest='baseCurrency',
        help="Symbol of the base currency.",
        )

# `symbols` option: Space separated list of stock symbols.
parser.add_argument('-s', '--symbols', type=str,
        default=None, dest='symbols',
        help="Space separated list of stock symbols to quote and cache.",
        )

# `date` option: Letting users select only records beyond this date.
parser.add_argument('--date', type=str,
        default=None, dest='date',
        help="Get quote on this date. Format YYYY-MM-DD.",
        )

args = parser.parse_args()

dbh = sqlite3.connect(args.db)
cursor = dbh.cursor()

# Capture present date. We will cache one quote per day as the whole
# utils package is not meant for realtime monitoring but rather for
# a one time use.
if args.date is None:
    date = datetime.date.today().strftime("%Y-%m-%d")
else:
    date = args.date

# collect stock symbols and currencies
if args.symbols is not None:
    # use symbols given from command line
    symbols = set()
    currencies = set()
    for s in args.symbols.split():
        m = re.search('^(\w+)=X$', s)
        if m:
            if args.baseCurrency == m[1][3:]:
                currencies.add(s = m[1][:3])
            continue

        symbols.add(s)

else:
    # use symbols in XFRS DB
    symbols = xfrs.getStocks(dbh) or list()
    if args.baseCurrency is not None:
        currencies = set(xfrs.getCurrencies(dbh) or list())


# Quote symbols
# -------------
quotes = dict()

# obtain cached quotes
for s, q in xfrs.getCachedQuote(dbh, list(symbols), date = date).items():
    q['status'] = 'cached'
    quotes[s] = q

# obtain online quotes
if len(symbols) > len(quotes.keys()):
    for s, q in xfrs.getOnlineQuote([s for s in symbols if s not in quotes], date = date).items():
        q['status'] = 'online'
        quotes[s] = q


# Quote currencies
# ----------------
if args.baseCurrency is not None:
    # add currencies of previously quoted stocks in the basket
    currencies.update([q['currency'] for q in quotes.values()])

    # filter out the base currency
    if args.baseCurrency in currencies: currencies.remove(args.baseCurrency)

    # map currencies to forex symbols
    fxMap = {c: c + args.baseCurrency + '=X' for c in currencies}

    # quote forex symbols
    fx = xfrs.getOnlineQuote(fxMap.values(), date = date)

    # recalculate the recorded stock quotes (to base currency)
    for s in quotes.keys():
        c = quotes[s]['currency']
        if c != args.baseCurrency and fxMap[c] in fx:
            quotes[s]['currency'] = args.baseCurrency
            quotes[s]['price'] *= fx[fxMap[c]]['price']

    # add forex quotes
    for c in currencies:
        if c not in quotes:
            quotes[c] = fx[fxMap[c]]
            quotes[c]['status'] = 'online'


# Report quotes
# -------------
for s, q in quotes.items():
    print(f'{s}\t{q["price"]:.3f} {q["currency"]} ({q.get("status","???")})')


# close DB
# --------
dbh.commit()
dbh.close()
